pkg/storage/unified/resource/index.go

```diff
@@ -320,44 +380,49 @@ func (i *Index) Search(ctx context.Context, request *SearchRequest) (*IndexResul
 	return &IndexResults{Values: results, Groups: groups}, nil
 }
 
-func (i *Index) Count() (uint64, error) {
-	var total uint64
+// Count returns the total doc count
+func (i *Index) Count() (int, error) {
+	total := 0
 	for _, shard := range i.shards {
 		count, err := shard.index.DocCount()
 		if err != nil {
-			i.log.Error("failed to get doc count", "error", err)
+			return 0, err
 		}
-		total += count
+		total += int(count)
 	}
 	return total, nil
 }
 
+// allTenants returns a list of all tenants in the index
 func (i *Index) allTenants() []string {
-	tenants := make([]string, 0, len(i.shards))
-	for tenant := range i.shards {
+	tenants := make([]string, 0)
+	for tenant, _ := range i.shards {
 		tenants = append(tenants, tenant)
 	}
 	return tenants
 }
 
-func (i *Index) getShard(tenant string) (Shard, error) {
+func (i *Index) getShard(tenant string) (*Shard, error) {
+	i.shardMutex.Lock()
```

Comment:
Now that we're building the index shards concurrently, we need to avoid concurrent map reads and writes